### PR TITLE
Add Yandex Music support

### PR DESCRIPTION
**Pull request by AgutinVBoy**
This is a new source manager for Yandex music. 

It supports songs, albums, playlists, artists and podcasts. Reqiures Token.

- `ymsearch:Let's Get It Started`
- https://music.yandex.ru/album/7091037/track/767978
- https://music.yandex.ru/album/7091037
- https://music.yandex.ru/users/music-blog/playlists/1103
- https://music.yandex.ru/artist/451

#### Yandex Music
```java
AudioPlayerManager playerManager = new DefaultAudioPlayerManager();

// create a new YandexMusicSourceManager with the token and register it
playerManager.registerSourceManager(new YandexMusicSourceManager("...", playerManager));
```

## Important information
Yandex Music is very location-dependent. You should either have a premium subscription or be located in one of the following countries:
- Azerbaijan
- Armenia
- Belarus
- Georgia
- Kazakhstan
- Kyrgyzstan
- Moldova
- Russia
- Tajikistan
- Turkmenistan
- Uzbekistan

Else you will only have access to podcasts.

## How to get token
1. (Optional) Open DevTools in your browser and on the Network tab enable trotlining.
2. Go to [https://oauth.yandex.ru/authorize?response_type=token&client_id=23cabbbdc6cd418abb4b39c32c41195d](https://oauth.yandex.ru/authorize?response_type=token&client_id=23cabbbdc6cd418abb4b39c32c41195d)
3. Authorize and grant access
4. The browser will redirect to the address like `https://music.yandex.ru/#access_token=AQAAAAAYc***&token_type=bearer&expires_in=31535645`. 
Very quickly there will be a redirect to another page, so you need to have time to copy the link. image
5. Your token, what is after `access_token`.

### CODE REVIEW

**Review by TopiSenpai — CHANGES_REQUESTED** [review]
I just skimmed it & will give it a full review later once you addressed most of this stuff
but here are some points I noticed:
- update the readme
- `YandexMusicAudioTrack#getTrackTitle` has no usage
- there is no point in passing `songId` that's just the identifier. Other sources use the isrc here which is something different
- why do you load the resulting mp3 like this? You can just create an `PersistentHttpStream` & skip the whole `loadItem` stuff(see deezer)
- `YandexMusicAudioTrack` should process an `Mp3AudioTrack` if it's an mp3 track
- `YanddexMusicSourceManager` does not need the `AudioPlayerManager`
- please cleanup the regex code(call matcher.find after each other not weirdly in each other)
- `json.get("bla") == null` will always be false. use `json.get("bla").isNull()`
- `token` in the config should be renamed to something more descriptive
- how long are tokens valid? do they ever expire?
- can you provide me with some example json responses for endpoints?
- I don't think you need the `YandexMusicTrackLoadingException`
- for html parsing use `jsoup` instead of `indexOf` & `substring`
- is that sign key guaranteed to not update? might be better to not hardcode it
- please orient yourself on the code styling I use(no ifs without `{}` as an example)
- shouldn't the music search prefix be `ymsearch:`?

besides those points above cool thing I guess. It seems pretty niche for most people

If you have any questions feel free to ask

**Comment by TopiSenpai** [comment]
> unfortunately, yandex music has no isrc. only IDs
then please remove it
> If I'm not mistaken, the token lives for one year. It cannot be refreshed automatically.
On the other hand, I heard that they live a very long time. Maybe lifetime.
please mention this in the docs
> I don't know :) I wrote yasearch because Yandex uses abbreviations for its products (Ya.Disk, Ya.Taxi, Ya.Music)
You use `Yandex Music` everywhere so I think it makes more sense
> Sure. Where can I send you responses? They are very large, so I don't think I should send them here. Maybe discord? (Eterna#9999)
you can just paste it as a file here or use https://hastebin.de

**Comment by AgutinVBoy** [comment]
> then please remove it

songId is required to get link to mp3 (YandexMusicAudioTrack#getDownloadURL). 

> please mention this in the docs

ok

> You use Yandex Music everywhere so I think it makes more sense

ok, I'll change it.

> responses

[track success](https://hastebin.de/lekerorigu.yaml), [track failed](https://hastebin.de/agetehaceb.yaml), [track failed (wrong country)](https://hastebin.de/bokepapagu.yaml), [playlist success](https://hastebin.de/lozawomewa.yaml), [playlist failed 404](https://hastebin.de/ukidolevad.yaml), [playlist failed (wrong country)](https://hastebin.de/ajunezalit.yaml).

If you need more, I'll do it.

**Comment by AgutinVBoy** [comment]
> you already pass the song id as identifier. there is no need to have it twice

Oh, I see. I forgot about that)

**Comment by AgutinVBoy** [comment]
About matcher. Is this okay? 
```java
var matcher = URL_PATTERN.matcher(reference.identifier);
var playlist_matcher = URL_PLAYLIST_PATTERN.matcher(reference.identifier);
if (playlist_matcher.find()) {
    matcher = playlist_matcher;
} else if (!matcher.find()) {
    return null;
}
```

**Comment by TopiSenpai** [comment]
why not
```java
var matcher = URL_PATTERN.matcher(reference.identifier);
if (matcher.find()) {
    // do stuff
    return
}

matcher = URL_PLAYLIST_PATTERN.matcher(reference.identifier);
if (matcher.find()) {
    // do stuff
    return
}
return null
```

File: main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicSourceManager.java [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+
+public class YandexMusicSourceManager implements AudioSourceManager, HttpConfigurable {
+    public static final Pattern URL_PATTERN = Pattern.compile("(https?://)?music\\.yandex\\.ru/(?<type1>artist|album)/(?<identifier>[0-9]+)/?((?<type2>track/)(?<identifier2>[0-9]+)/?)?");
+    public static final Pattern URL_PLAYLIST_PATTERN = Pattern.compile("(https?://)?music\\.yandex\\.ru/(?<type1>users)/(?<identifier>[0-9A-Za-z@.-]+)/playlists/(?<identifier2>[0-9]+)/?");
```
Comment (TopiSenpai) [inline main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicSourceManager.java:35]:
```suggestion
    public static final Pattern URL_PLAYLIST_PATTERN = Pattern.compile("(https?://)?music\\.yandex\\.ru/users/(?<identifier>[0-9A-Za-z@.-]+)/playlists/(?<identifier2>[0-9]+)/?");
```

File: main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicSourceManager.java [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+            return AudioReference.NO_TRACK;
+        var tracks = this.parseAlbum(json.get("result").get("volumes"));
+        if (tracks.isEmpty())
+            return AudioReference.NO_TRACK;
+        return new BasicAudioPlaylist(json.get("result").get("title").text(), tracks, null, false);
+    }
+
+    private AudioItem getTrack(String id) throws IOException {
+        var json = this.getJson(PUBLIC_API_BASE + "/tracks/" + id);
+        if (json.isNull())
```
Comment (TopiSenpai) [inline main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicSourceManager.java:114]: you can combine these ifs together

File: main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+- Georgia
+- Kazakhstan
+- Kyrgyzstan
+- Moldova
+- Russia
+- Tajikistan
+- Turkmenistan
+- Uzbekistan
+
+Else you will only have access to podcasts.
```
Comment (TopiSenpai) [inline main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/README.md:25]: I'd rather have this in the root readme
you can hide some stuff with a details html block

File: README.md [hunk 4]
```diff
@@ -6,7 +6,7 @@ A collection of additional [Lavaplayer](https://github.com/sedmelluq/lavaplayer)
 * [Spotify*](https://www.spotify.com) playlists/albums/songs/artists(top tracks)/search results
 * [Apple Music*](https://www.apple.com/apple-music/) playlists/albums/songs/artists/search results(Big thx to [ryan5453](https://github.com/ryan5453) for helping me)
 * [Deezer](https://www.deezer.com) playlists/albums/songs/artists/search results(Big thx to [ryan5453](https://github.com/ryan5453) and [melike2d](https://github.com/melike2d) for helping me)
-* [Yandex Music](https://music.yandex.ru) playlists/albums/songs/artists/podcasts/search results
+* [Yandex Music](https://music.yandex.ru) playlists/albums/songs/artists/podcasts/search results(Big thx to [AgutinVBoy](https://github.com/agutinvboy) for helping me)
```
Comment (TopiSenpai) [inline README.md:9]:
```suggestion
* [Yandex Music](https://music.yandex.ru) playlists/albums/songs/artists/podcasts/search results(Thx to [AgutinVBoy](https://github.com/agutinvboy) for implementing it)
```

File: main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicSourceManager.java [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        request.setHeader("Accept", "application/json");
+        request.setHeader("Authorization", "OAuth " + this.accessToken);
+        return HttpClientTools.fetchResponseAsJson(this.httpInterfaceManager.getInterface(), request);
+    }
+
+    public String getDownloadStrings(String uri) throws IOException {
+        var request = new HttpGet(uri);
+        request.setHeader("Accept", "application/json");
+        request.setHeader("Authorization", "OAuth " + this.accessToken);
+        return HttpClientTools.fetchResponseLines(this.httpInterfaceManager.getInterface(), request, "UTF-8")[0];
```
[the remark this answers is not in the document]
Comment (AgutinVBoy) [reply main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicSourceManager.java]: in lower case? If so, then it's resolved

[the remark this answers is not in the document]
Comment (TopiSenpai) [reply main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicSourceManager.java]: not exactly

> looking at the lavaplayer code this actually something else

> name – Name of the operation to include in exception messages.

[the remark this answers is not in the document]
Comment (TopiSenpai) [reply main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicSourceManager.java]: As an example
Just something descriptive

[the remark this answers is not in the document]
Comment (AgutinVBoy) [reply main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicSourceManager.java]: done

File: main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicAudioTrack.java [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    private String getDownloadURL(String id) throws IOException, NoSuchAlgorithmException {
+        var json = this.sourceManager.getJson(YandexMusicSourceManager.PUBLIC_API_BASE + "/tracks/" + id + "/download-info");
+        if (json.isNull()|| json.get("result").values().isEmpty()) {
+            throw new FriendlyException("No download URL found for track " + id, FriendlyException.Severity.COMMON, null);
+        }
+
+        var downloadInfoLink = json.get("result").values().get(0).get("downloadInfoUrl").text();
+        String downloadInfo = this.sourceManager.getDownloadStrings(downloadInfoLink);
+        if (downloadInfo == null) {
+            throw new FriendlyException("No download URL found for track " + id, FriendlyException.Severity.COMMON, null);
```
Comment (TopiSenpai) [inline main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicAudioTrack.java:58]: you should throw `IllegalStateException` here

File: main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicAudioTrack.java [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+
+    @Override
+    protected AudioTrack makeShallowClone() {
+        return new YandexMusicAudioTrack(this.trackInfo, this.artworkURL, this.sourceManager);
+    }
+
+    private String getDownloadURL(String id) throws IOException, NoSuchAlgorithmException {
+        var json = this.sourceManager.getJson(YandexMusicSourceManager.PUBLIC_API_BASE + "/tracks/" + id + "/download-info");
+        if (json.isNull()|| json.get("result").values().isEmpty()) {
+            throw new FriendlyException("No download URL found for track " + id, FriendlyException.Severity.COMMON, null);
```
Comment (TopiSenpai) [inline main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicAudioTrack.java:52]: you should throw `IllegalStateException` here

File: main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicAudioTrack.java [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+        var downloadInfoLink = json.get("result").values().get(0).get("downloadInfoUrl").text();
+        String downloadInfo = this.sourceManager.getDownloadStrings(downloadInfoLink);
+        if (downloadInfo == null) {
+            throw new FriendlyException("No download URL found for track " + id, FriendlyException.Severity.COMMON, null);
+        }
+
+        Document doc = Jsoup.parse(downloadInfo, "", Parser.xmlParser());
```
Comment (TopiSenpai) [inline main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicAudioTrack.java:61]: var

File: main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicAudioTrack.java [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        var host = doc.select("host").text();
+        var path = doc.select("path").text();
+        var ts = doc.select("ts").text();
+        var s = doc.select("s").text();
+
+        var sign = "XGRlBW9FXlekgbPrRHuSiA" + path + s;
+        var md = MessageDigest.getInstance("MD5");
+        var digest = md.digest(sign.getBytes(StandardCharsets.UTF_8));
+        var sb = new StringBuilder();
+        for (byte b : digest)
```
Comment (TopiSenpai) [inline main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicAudioTrack.java:71]: missing `{}`

File: main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicAudioTrack.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+        }
+
+        Document doc = Jsoup.parse(downloadInfo, "", Parser.xmlParser());
+        var host = doc.select("host").text();
+        var path = doc.select("path").text();
+        var ts = doc.select("ts").text();
+        var s = doc.select("s").text();
+
+        var sign = "XGRlBW9FXlekgbPrRHuSiA" + path + s;
+        var md = MessageDigest.getInstance("MD5");
```
Comment (TopiSenpai) [inline main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicAudioTrack.java:68]: can we reuse the md5 instance?

**Review by TopiSenpai — CHANGES_REQUESTED** [review]

File: main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicAudioTrack.java (hunk 10 above)
Comment (AgutinVBoy) [reply main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicAudioTrack.java:68]: I think so

Comment (AgutinVBoy) [reply main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicAudioTrack.java:66]: Do you think it should be written elsewhere?

Comment (TopiSenpai) [reply main/src/main/java/com/github/topisenpai/lavasrc/yandexmusic/YandexMusicAudioTrack.java:66]: It doesn't really matter
Ignore I said anything :^)

**Comment by TopiSenpai** [comment]
I can't really test this as I don't have an account there and tbh I don't want to create one either
Do you have a bot/node running with this or want to give me an access token?
Either works for me I guess

**Comment by TopiSenpai** [comment]
code-wise it looks fine at least

**Comment by AgutinVBoy** [comment]
Sure. You can join in my [discord server](https://discord.gg/FA9d7MQk) 
Prefixes `>>` or `/2`

**Review by TopiSenpai — CHANGES_REQUESTED** [review]
looks fine to me now
how does it behave with big playlists?
does it load the full or does it need pagination?

**Review by TopiSenpai — APPROVED** [review]
nice work

**Pull request merged**